tests/test_preprocess_model_definition.py

```diff
@@ -0,0 +1,223 @@
+            """
+        )
+        with pytest.raises(ModelError) as excinfo:
+            build_model(override_dict=override, scenario="scenario_1")
+
+        assert check_error_or_warning(
+            excinfo, "(scenarios, scenario_1) | Unrecognised override name: foo."
+        )
+
+
+class TestTemplateSolver:
+    @pytest.fixture
+    def dummy_solved_template(self) -> TemplateSolver:
+        text = """
+        templates:
+            T1:
+                A: ["foo", "bar"]
+                B: 1
+            T2:
+                C: bar
+                template: T1
+            T3:
+                template: T1
+                B: 11
+            T4:
+                template: T3
+                A: ["bar", "foobar"]
+                B: "1"
+                C: {"foo": "bar"}
+                D: true
+        a:
+            template: T1
+            a1: 1
+        b:
+            template: T3
+        c:
+            template: T4
+            D: false
+        """
+        yaml_data = read_rich_yaml(text)
+        return TemplateSolver(yaml_data)
+
+    def test_inheritance_templates(self, dummy_solved_template):
+        templates = dummy_solved_template.resolved_templates
+        assert all(
+            [
+                templates.T1 == {"A": ["foo", "bar"], "B": 1},
+                templates.T2 == {"A": ["foo", "bar"], "B": 1, "C": "bar"},
+                templates.T3 == {"A": ["foo", "bar"], "B": 11},
+                templates.T4
+                == {"A": ["bar", "foobar"], "B": "1", "C": {"foo": "bar"}, "D": True},
+            ]
+        )
+
+    def test_template_inheritance_data(self, dummy_solved_template):
+        data = dummy_solved_template.resolved_data
+        assert all(
+            [
+                data.a == {"A": ["foo", "bar"], "B": 1, "a1": 1},
+                data.b == {"A": ["foo", "bar"], "B": 11},
+                data.c
+                == {"A": ["bar", "foobar"], "B": "1", "C": {"foo": "bar"}, "D": False},
+            ]
+        )
+
+    def test_invalid_template_error(self):
+        text = read_rich_yaml(
+            """
+            templates:
+                T1: "not_a_yaml_block"
+                T2:
+                    foo: bar
+            a:
+                template: T2
+            """
+        )
+        with pytest.raises(
+            ModelError, match="Template definitions must be YAML blocks."
+        ):
+            TemplateSolver(text)
+
+    def test_circular_template_error(self):
+        text = read_rich_yaml(
+            """
+            templates:
+                T1:
+                    template: T2
+                    bar: foo
+                T2:
+                    template: T1
+                    foo: bar
+            a:
+                template: T2
+            """
+        )
+        with pytest.raises(ModelError, match="Circular template reference detected"):
+            TemplateSolver(text)
+
+    def test_incorrect_template_placement_error(self):
```

Comment:
should we be confirming that `template` is not defined in `config`, or is that not making its way into the template solver?

Another place where `template` might now be ending up is in top-level params (the `parameters` key). We should ensure that this is working as expected.

Reply:
Both of these are fine. We should see template as something a user decides they want to re-use.
Although using it in `config` and `parameters` would be weird, it's ok as long as the final shape fits our expected schema.

This is the advantage of moving this to the IO: our job is to ensure that what is fed into calliope is good, not to police how users should / should not define their files.

Additionally, once we move the parameter definition from the schema to the model YAML, templates _would_ be useful in that section.

Reply:
I will add this check to the 'overrides' section, though. Since it can alter templates due to our ordering (overrides > templates).
An override should not make use of `template:`, otherwise it would result in weird behaviour.